cpp/include/kvikio/bounce_buffer.hpp

```diff
@@ -11,48 +11,153 @@
 namespace kvikio {
 
 /**
- * @brief Singleton class to retain host memory allocations
+ * @brief Allocator for page-aligned host memory
  *
- * Call `AllocRetain::get` to get an allocation that will be retained when it
+ * Uses std::aligned_alloc to allocate host memory aligned to page boundaries (typically 4096
+ * bytes). This allocator is suitable for Direct I/O operations that require page-aligned buffers
+ * but do not need CUDA context (i.e., host-to-host transfers only).
+ */
+class PageAlignedAllocator {
+ public:
+  /**
+   * @brief Allocate page-aligned host memory
+   *
+   * @param size Requested size in bytes (will be rounded up to page boundary)
+   * @return Pointer to allocated memory
+   */
+  void* allocate(std::size_t size);
+
+  /**
+   * @brief Deallocate memory previously allocated by this allocator
+   *
+   * @param buffer Pointer to memory to deallocate
+   * @param size Size of the allocation (unused, for interface consistency)
+   */
+  void deallocate(void* buffer, std::size_t size);
+};
+
+/**
+ * @brief Allocator for CUDA pinned host memory
+ *
+ * Uses cudaMemHostAlloc to allocate pinned (page-locked) host memory that can be efficiently
+ * transferred to/from GPU device memory. The allocation is only guaranteed to be aligned to "at
+ * least 256 bytes". It is NOT guaranteed to be page aligned.
+ *
+ * @note Do NOT use with Direct I/O - lacks page alignment guarantee
+ */
+class CudaPinnedAllocator {
+ public:
+  /**
+   * @brief Allocate CUDA pinned host memory
+   *
+   * @param size Requested size in bytes
+   * @return Pointer to allocated pinned memory
+   */
+  void* allocate(std::size_t size);
+
+  /**
+   * @brief Deallocate memory previously allocated by this allocator
+   *
+   * @param buffer Pointer to memory to deallocate
+   * @param size Size of the allocation (unused, for interface consistency)
+   */
+  void deallocate(void* buffer, std::size_t size);
+};
+
+/**
+ * @brief Allocator for page-aligned AND CUDA-registered pinned host memory
+ *
+ * Combines the benefits of both page alignment (for Direct I/O) and CUDA registration
+ * (for efficient host-device transfers). Uses std::aligned_alloc followed by
+ * cudaMemHostRegister to achieve both properties.
+ *
+ * @note This is the required allocator for Direct I/O with device memory. Requires a valid CUDA
+ * context when allocating.
+ */
+class CudaPageAlignedPinnedAllocator {
```

Comment:
This combinatorial set of types reminds of the design from "Modern C++ Design", where each of these options would be a separate template parameter (or something like that). Probably not required here since there are only three types, but wanted to leave a note for (future) consideration.